Don't crash on dead code in ICodeReader

lrytz opened the pull request:
Under `-Ybackend:GenBCode`, the scala compiler occasionally emits dead code, see discussion on PR #3726 for example.

Such a class file crashes ICodeReader. This was uncovered because under `-optimise`, the compiler switches to the ICode-GenASM pipeline, even if the `-Ybackend:GenBCode` flag is present.

There are multiple [tests that fail before](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1227/consoleFull) basically because the compiler emits inliner warnings ("could not inline") when the ICodeReader failed.

lrytz committed:
Don't crash on dead code in ICodeReader

retronym commented:
Maybe the test could mutate a valid class with ASM to inject dead frames.  That way the regular compiler is still in charge of signatures. 

Does icodereader also work for java classes? That would be simpler to craft in the test.

lrytz commented:
I don't think we ever inline a java defined method.

Maybe it could work with an interpreter test, and we can change a class in memory, as you suggest.

lrytz committed:
Tone down the noise in the output of t7974

This makes it pass under -Ybackend:GenBCode

lrytz committed:
Make test repl-javap-app independent of precise constant pool indices

Allows the test to pass on both backends

som-snytt commented on test/files/run/repl-javap-app.scala:
```diff
@@ -7,4 +7,15 @@ object MyApp extends App {
+  override def normalize(s: String) = s match {
+    case hasConstantPoolRef(start, ref, end) => start + "#XX" + end
+    case _ => super.normalize(s)
+  }
```
Nice use of hook.

lrytz committed:
Test: classfile reading during inlining works if there is dead code

lrytz commented:
Fixed failing tests, added a test for the fix in ICodeReader. review by @retronym 
`-Ybackend:GenBCode` build [passed](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1229/consoleFull)
cc @dragos

retronym commented:
LGTM. I like the test.
